File: Project_1_Java/src/PLD/classes/Parser.java
package PLD.classes;

import java.util.*;

public class Parser {

    private Lexer lexer;
    private Stack<Token> operators = new Stack<>();
    private Queue<Token> output = new LinkedList<>();
    private Token lookahead;

    public Parser(Lexer lexer) {
        this.lexer = lexer;
    }

    /**
     * This functions consumes the next token from our Token Table
     */
    public void consumeToken() {
        this.lookahead = this.lexer.getToken();
    }

    /***
     * Function to convert the Infix Regex notation to a postfix one
     */
    public String parse() {
        for (int i = 0; i < this.lexer.getRegexLength(); i++) {
            consumeToken();
            if(this.lookahead.getId().equals("OPEN_PAR")){
                this.operators.push(this.lookahead);
            }
            else if(this.lookahead.getId().equals("CLOSE_PAR")){
                while (!this.operators.peek().getId().equals("OPEN_PAR")){
                    this.output.add(this.operators.pop());
                }
                this.operators.pop();
            }else if(this.lookahead.getId().equals("CHAR")){
                this.output.add(this.lookahead);
            }else {
                while (this.operators.size() > 0 && this.lookahead.getPrecedence() <= this.operators.peek().getPrecedence()
                && !this.operators.peek().getId().equals("OPEN_PAR") && !this.operators.peek().getId().equals("CLOSE_PAR")){
                    this.output.add(this.operators.pop());
                }
                this.operators.push(this.lookahead);
            }
        }

        int operatorsSize = this.operators.size();
        for (int i = 0; i < operatorsSize; i++) {
            Token token = this.operators.pop();
            this.output.add(token);
        }

        this.operators.clear();

        StringBuilder postfixRegex = new StringBuilder();

        for (Token t : output) {
            postfixRegex.append(t.getValue());
        }

        return postfixRegex.toString();
    }
}

File: Project_2/classes/Parser.py
from classes.Token import Token


class Parser:
    def __init__(self, lexer):
        self.lexer = lexer
        self.operators = []
        self.output = []
        self.lookahead = None

    def consume_token(self):
        self.lookahead = self.lexer.get_token()

    def parse(self):
        for i in range(len(self.lexer.regex)):
            self.consume_token()
            if self.lookahead.id == 'OPEN_PAR':
                self.operators.append(self.lookahead)
            elif self.lookahead.id == 'CLOSE_PAR':
                while self.operators and self.operators[-1].id != 'OPEN_PAR':
                    self.output.append(self.operators.pop())
                if self.operators:
                    self.operators.pop()
            elif self.lookahead.id == 'CHAR':
                self.output.append(self.lookahead)
            else:
                while len(self.operators) > 0 and self.lookahead.precedence <= self.operators[-1].precedence \
                        and self.operators[-1].id != 'OPEN_PAR' and self.operators[-1].id != 'CLOSE_PAR':
                    self.output.append(self.operators.pop())
                self.operators.append(self.lookahead)

        operators_size = len(self.operators)
        for i in range(operators_size):
            token = self.operators.pop()
            self.output.append(token)
        self.operators.clear()

        postfix_regex = ''

        for t in self.output:
            postfix_regex += t.value

        return postfix_regex
